Source/Scene/ShadowMap.js

```diff
@@ -0,0 +1,397 @@
+        if (!defined(shadowMap.framebuffer) || textureChanged) {
+            createFramebuffers(shadowMap, context);
+        }
+    }
+
+    ShadowMap.createShadowCastProgram = function(shaderProgram) {
+        // TODO : unused right now
+        // TODO : vertex shader won't be optimized
+        // TODO : handle mismatched varyings
+        var vs = shaderProgram.vertexShaderText;
+        var fs =
+            'void main()\n' +
+            '{\n' +
+            '    gl_FragColor = vec4(0.0);\n' +
+            '}\n';
+
+        return ShaderProgram.fromCache({
+            vertexShaderSource : vs,
+            fragmentShaderSource : fs,
+            attributeLocations : shaderProgram._attributeLocations
+        });
+    };
+
+    ShadowMap.createReceiveShadowsVertexShader = function(vs) {
+        vs = ShaderSource.replaceMain(vs, 'czm_shadow_main');
+        vs +=
+            'varying vec3 czm_shadowMapCoordinate; \n' +
+            'void main() \n' +
+            '{ \n' +
+            '    czm_shadow_main(); \n' +
+            '    czm_shadowMapCoordinate = (czm_shadowMapMatrix * gl_Position).xyz; \n' +
+            '} \n';
+        return vs;
+    };
+
+    ShadowMap.createReceiveShadowsFragmentShader = function(fs) {
+        fs = ShaderSource.replaceMain(fs, 'czm_shadow_main');
+        fs +=
+            'varying vec3 czm_shadowMapCoordinate; \n' +
+            'void main() \n' +
+            '{ \n' +
+            '    czm_shadow_main(); \n' +
+            '    float depth = czm_shadowMapCoordinate.z; \n' +
+            '    float shadowDepth = texture2D(czm_shadowMapTexture, czm_shadowMapCoordinate.xy).r; \n' +
+            '    // TODO : remove if \n' +
+            '    if (depth - 0.005 > shadowDepth) { \n' +
+            '        gl_FragColor.rgb *= 0.2; \n' +
+            '    } \n' +
+            '} \n';
+        return fs;
+    };
+
+    ShadowMap.prototype.setSize = function(size) {
+        this._size = size;
+        this.renderState.viewport.width = this._size;
+        this.renderState.viewport.height = this._size;
+    };
+
+    function updateDebugShadowViewCommand(shadows, frameState) {
+        // Draws the shadow map on the bottom-right corner of the screen
+        var screenWidth = frameState.context.drawingBufferWidth;
+        var screenHeight = frameState.context.drawingBufferHeight;
+        var size = Math.min(screenWidth, screenHeight) * 0.3;
+
+        var x = screenWidth - size;
+        var y = 0;
+        var width = size;
+        var height = size;
+
+        if (!defined(shadows._debugShadowViewCommand)) {
+            var fs =
+                'varying vec2 v_textureCoordinates; \n' +
+                'void main() \n' +
+                '{ \n' +
+                '    float shadow = texture2D(czm_shadowMapTexture, v_textureCoordinates).r; \n' +
+                '    gl_FragColor = vec4(shadow, shadow, shadow, 1.0); \n' +
```

Comment:
Up to you, but I think `vec4(vec3(shadow), 1.0);` is better style.  Don't you remember that 565 lecture!??!?!!??!